core/src/main/scala/com/stripe/dagon/Literal.scala

```diff
@@ -10,15 +10,68 @@ sealed trait Literal[N[_], T] {
 
 object Literal {
 
-  case class Const[N[_], T](override val evaluate: N[T]) extends Literal[N, T]
+  case class Const[N[_], T](override val evaluate: N[T]) extends Literal[N, T] {
+    // cache hashCode since we hit it so much.
+    override val hashCode = (getClass, evaluate).hashCode
 
-  case class Unary[N[_], T1, T2](arg: Literal[N, T1], fn: N[T1] => N[T2]) extends Literal[N, T2]
+    // We often will cache the N[T] -> Literal[N, T] mapping, so we
+    // deal with referentially equal things, which can give cheap equality
+    override def equals(that: Any) =
+      that match {
+        case thatAR: AnyRef if thatAR eq this => true
```

Comment:
It seems like the reference identity check could be factored out into `Literal` to clean up this code a bit?

Reply:
How were you thinking? Doesn't seem like I can actually get shorter code, and I also worry slightly about worse preformance.